File: hwserver2/db/device.rb
class DB::Device < DB
  def initialize hwid, db
    @db = db
    @hwid = hwid
    @record = db[:devices].find(hwid: hwid).first
  end

  def algorithm
    record = @db[:algorithms].find(:'algorithm-id' => @record['algorithm-id']).first
    content(record, 'algorithm') if record
  end

  def interface
    record = @db[:interfaces].find(:'interface-id' => @record['interface-id']).first
    content(record, 'interface') if record
  end

  def manual?
    @record['manual']
  end

  def group
    @record['group']
  end

  def group?
    !(@record['group'].empty? || @record['group'] == nil)
  end
end

File: web/app/routes/devices/device.js
import Ember from 'ember';
import authRouteMixin from '../../mixins/auth-route';

export default Ember.Route.extend(authRouteMixin, {
  controllerName: 'device',

  setupController: function(controller, model) {
    var algorithms = this.store.findAll('algorithm');
    controller.set('algorithms', algorithms);
    if(model) {
      controller.set('devicesController.currentDeviceId', model.get('id'));
    }
    this._super(controller, model);
  },

  model(params) {
    if(params['device_id'] === 'new') {
      this.get('store').createRecord('device', {hwid: Math.round(Math.random()*100000).toString()}).save().then(function(device) {
        this.transitionTo('/devices/' + device.get('id'));
      }.bind(this));
    } else {
      var device = this.store.findRecord('device', params['device_id']);

      return device.then(function(device) {
        if(!device.get('build.content')) {
          var build = this.store.createRecord('build');
          build.save().then(function(build) {
            device.set('build', build);
            device.save();
          });
        }
        return device;
      }.bind(this), function(device) {
        this.transitionTo('/devices');
      }.bind(this));
    }
  }
});
